### PR TITLE
Add support for short number and emegerncy number in libPhoneNumber-iOS

### PR DESCRIPTION
**Pull request by paween**
This is a port of ShortNumberInfo class from Java version of libPhoneNumber into a category on NBPhoneNumberUtil class.

New short number metadata is included in this commit, but is not compiled in by default. In order to add short number/emergency number support, one can define SHORT_NUMBER_SUPPORT macro in the build settings.

### CODE REVIEW

File: libPhoneNumber/Internal/NBRegExMatcher.m [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+  NSRange wholeStringRange = NSMakeRange(0, string.length);
+
+  // Prefix match (lookingAt()) search
+  NSRegularExpression *prefixRegEx =
+      [[NBRegularExpressionCache sharedInstance] regularExpressionForPattern:nationalNumberPattern];
+  NSTextCheckingResult *prefixResult = [prefixRegEx firstMatchInString:string
+                                                               options:NSMatchingAnchored
+                                                                 range:wholeStringRange];
+  if (prefixResult.numberOfRanges < 1) {
```
Comment (LowAmmo) [inline libPhoneNumber/Internal/NBRegExMatcher.m]: Might be cleaner to do `<= 0`

File: libPhoneNumber/Internal/NBRegExMatcher.m [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    // No prefix match found.
+    return NO;
+  } else {
```
Comment (LowAmmo) [inline libPhoneNumber/Internal/NBRegExMatcher.m]: Don't really need the 'else' here (since the above if block returns)

File: libPhoneNumber/Internal/NBRegExMatcher.m [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    // Found prefix match, but need to see if exact match works as well.
+    // Exact match (matches()) search.
+    NSTextCheckingResult *exactResult = [regEx firstMatchInString:string
+                                                     options:NSMatchingAnchored
+                                                       range:wholeStringRange];
+
+    if (exactResult.numberOfRanges > 0) {
```
Comment (LowAmmo) [inline libPhoneNumber/Internal/NBRegExMatcher.m]: Could make a one-liner - `return (allowsPrefixMatch || exactResult.numberOfRanges > 0)`

File: libPhoneNumber/Internal/NBRegularExpressionCache.m [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    NSRegularExpression *cachedObject = [self.cache objectForKey:pattern];
+    if (cachedObject != nil) {
+      return cachedObject;
+    }
+
+    NSError *error = nil;
+    NSRegularExpression *regEx = [[NSRegularExpression alloc] initWithPattern:pattern
+                                                                      options:kNilOptions
+                                                                        error:&error];
+    if (regEx == nil && error != nil) {
```
Comment (LowAmmo) [inline libPhoneNumber/Internal/NBRegularExpressionCache.m]: Probably need to make this an ||... Otherwise if for some reason regEx is nil, and error is nil, it will still continue on.

File: libPhoneNumber/Internal/NBRegularExpressionCache.m [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+      NSLog(@"An error parsing a regular expression: %@", error);
```
Comment (LowAmmo) [inline libPhoneNumber/Internal/NBRegularExpressionCache.m]: Then you can flex how you handle the error, to account for the possibility of it being nil.

File: libPhoneNumberTests/NBPhoneNumberUtil+ShortNumberTest.h [hunk 6]
```diff
@@ -0,0 +1,29 @@
+//
+//  NBPhoneNumberUtil+ShortNumberTest.h
```
Comment (LowAmmo) [inline libPhoneNumberTests/NBPhoneNumberUtil+ShortNumberTest.h]: Rename, possibly...  Seeing "Test" at the end of a class name usually implies it's an actual test case (ie. extends XCTest)

File: libPhoneNumberTests/generatedJSON/ShortNumberMetadata.json [hunk 7]
```diff
@@ -0,0 +1 @@
...0",null,null,[3,4,5,6]],[null,null,"11[67][0-2]\\d{3}|56161561",null,null,null,"1160530",null,null,[7,8]],null,null,null,"IN",null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,[null,null,"1(?:0[0128]|12|298)|2611",null,null,null,"108",null,null,[3,4]],null,[null,null,"1(?:0(?:[01248]|5(?:010|6|902)|7(?:[07]|80)|9[0157])|1(?:[289]|[67][0-2]\\d{3})|2(?:1|98)|39|4[01]|55[23]\\d{2}|7000|9(?:0[019]|47|50|6[1347]|[89]))|2611|5(?:14(?:2[5-9]|[34]\\d)|3000|6161(?:17[89]|561)|757555)",null,null,null,"108"],[null,null,"5(?:14(?:2[5-9]|[34]\\d)|757555)",null,null,null,"5757555",null,null,[5,7]],[null,null,"1(?:1(?:[67][0-2]\\d{3}|[89])|21|4[01]|55330|7\\d{3}|9(?:[89]|09))|5(?:3000|6161(?:17[89]|561))",null,null,null,"53000"],[null,null,"1(?:39|90[019])|5(?:14(?:2[5-9]|[34]\\d)|6161(?:17[89]|561)|757555)",null,null,null,"51431",null,null,[3,4,5,7,8]]],"IQ":[null,[null,null,"[1479]\\d{2,4}",null,null,null,null,null,null,[3,4,5]],null,null,[null,null,null,null,null,null,null,null,null,[-1]],[null,null,null,null,null,null,null,null,null,[-1]],null,null,null,"IQ",null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,[null,null,"1(?:
```
Comment (LowAmmo) [inline libPhoneNumberTests/generatedJSON/ShortNumberMetadata.json]: Would it be possible to format this into a more readable structure?

https://jsonformatter.org/

File: libPhoneNumber/Internal/NBRegExMatcher.m (hunk 1 above)
Comment (paween) [reply libPhoneNumber/Internal/NBRegExMatcher.m]: Done

File: libPhoneNumberTests/generatedJSON/ShortNumberMetadata.json (hunk 7 above)
Comment (paween) [reply libPhoneNumberTests/generatedJSON/ShortNumberMetadata.json]: This would be possible but it wouldn't really be readable anyway. The JSON is generated by libPhoneNumber build tool and it only contains data with pretty much no metadata (as in, it doesn't tell you which field is which), so all you're going to get is something like this which is not really readable :(

```
{  
  "AC":[  
    null,
    [  
      null,
      null,
      "9\\d{2}",
      null,
      null,
      null,
      null,
      null,
      null,
      [  
        3
      ]
    ],
    null,
    null,
    [  
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      [  
        -1
      ]
    ],
    [  
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      [  
        -1
      ]
    ],
    null,
    null,
    null,
    "AC",
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    [  
      null,
      null,
      "911",
      null,
      null,
      null,
      "911"
    ],
    null,
    [  
      null,
      null,
      "911",
      null,
      null,
      null,
      "911"
    ],
    [  
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      [  
        -1
      ]
    ],
    [  
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      [  
        -1
      ]
    ],
    [  
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      [  
        -1
      ]
    ]
  ]
}
```

File: libPhoneNumber/Internal/NBRegExMatcher.m (hunk 2 above)
Comment (paween) [reply libPhoneNumber/Internal/NBRegExMatcher.m]: I prefer having the else, to indicate that this is the "else" case of the condition above. If you feel strongly about it, let me know.

File: libPhoneNumberTests/NBPhoneNumberUtil+ShortNumberTest.h (hunk 6 above)
Comment (paween) [reply libPhoneNumberTests/NBPhoneNumberUtil+ShortNumberTest.h]: Renamed it to NBPhoneNumberUtil+ShortNumberTestHelper

File: libPhoneNumber/Internal/NBRegExMatcher.m (hunk 3 above)
Comment (paween) [reply libPhoneNumber/Internal/NBRegExMatcher.m]: Done

File: libPhoneNumber/Internal/NBRegularExpressionCache.m (hunk 4 above)
Comment (paween) [reply libPhoneNumber/Internal/NBRegularExpressionCache.m]: Done

File: .travis.yml [hunk 8]
```diff
@@ -4,6 +4,8 @@ before_install:
 - gem install slather -N 
 
 script:
-- xctool -project libPhoneNumber.xcodeproj -scheme libPhoneNumberiOSTests -sdk iphonesimulator clean build test ONLY_ACTIVE_ARCH=NO GCC_INSTRUMENT_PROGRAM_FLOW_ARCS=YES GCC_GENERATE_TEST_COVERAGE_FILES=YES  
+- xcodebuild -project libPhoneNumber.xcodeproj -scheme libPhoneNumberiOSTests clean
```
Comment (dmaclach) [inline .travis.yml]: Can we break this out into a separate PR (given that it has nothing to do with the short number and emergency number)

File: libPhoneNumber/Internal/NBRegularExpressionCache.m [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    if (cachedObject != nil) {
+      return cachedObject;
+    }
+
+    NSError *error = nil;
+    NSRegularExpression *regEx = [[NSRegularExpression alloc] initWithPattern:pattern
+                                                                      options:kNilOptions
+                                                                        error:&error];
+    if (regEx == nil || error != nil) {
+      NSLog(@"An error parsing a regular expression: %@", error);
```
Comment (dmaclach) [inline libPhoneNumber/Internal/NBRegularExpressionCache.m]: Should error be returned as part of the interface? NSLog is not a useful interface for most users.

File: libPhoneNumber/Internal/NBRegularExpressionCache.m [hunk 10]
```diff
@@ -0,0 +1,60 @@
+//
+//  NBRegularExpressionCache.m
```
Comment (dmaclach) [inline libPhoneNumber/Internal/NBRegularExpressionCache.m]: This part of the change should be broken out to a separate PR because it has nothing to do with the short number support.

File: libPhoneNumber/Internal/NBRegExMatcher.h [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+//  libPhoneNumber
+//
+//  Created by Paween Itthipalkul on 11/29/17.
+//  Copyright © 2017 Google LLC. All rights reserved.
+//
+
+#import <Foundation/Foundation.h>
+
+@class NBPhoneNumberDesc;
+
```
Comment (dmaclach) [inline libPhoneNumber/Internal/NBRegExMatcher.h]: Why are we pulling out what is essentially a single method into its own class?

File: libPhoneNumber/Internal/NBPhoneNumberUtil+Category.h [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+#import <Foundation/Foundation.h>
+
+#import "NBPhoneNumberUtil.h"
+
+@class NBMetadataHelper, NBRegExMatcher;
+
+/**
+ This interface exposes properties used in NBPhoneNumberUtil categories.
+ */
+@interface NBPhoneNumberUtil()
```
Comment (dmaclach) [inline libPhoneNumber/Internal/NBPhoneNumberUtil+Category.h]: normally we don't expose a class extension in a header.

File: libPhoneNumber/NBPhoneNumberUtil+ShortNumber.m [hunk 13]
```diff
@@ -0,0 +1,323 @@
...
+
+static NSString * const PLUS_CHARS_PATTERN = @"[+\uFF0B]+";
+
+@interface NBPhoneNumberUtil()
+
+@property (nonatomic) NSDictionary<NSNumber *, NSArray<NSString *> *> *countryToRegionCodeMap;
+
+@end
+
+@implementation NBPhoneNumberUtil (ShortNumber)
```
Comment (dmaclach) [inline libPhoneNumber/NBPhoneNumberUtil+ShortNumber.m]: Splitting this out to a separate code file is just making things more confusing IMHO. Could it not be in the NBPhoneNumberUtil file?

File: libPhoneNumberTests/NBPhoneNumberUtil+ShortNumberTestHelper.m [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+#import "NBPhoneNumberUtil+Category.h"
+
+#if SHORT_NUMBER_SUPPORT
+
+@implementation NBPhoneNumberUtil(ShortNumberTestHelper)
+
+- (NSString *)exampleShortNumberForCost:(NBEShortNumberCost)cost regionCode:(NSString *)regionCode {
+  NBPhoneMetaData *metadata = [self.helper shortNumberMetadataForRegion:regionCode];
+  if (metadata == nil) {
+    return @"";
```
Comment (dmaclach) [inline libPhoneNumberTests/NBPhoneNumberUtil+ShortNumberTestHelper.m]: is returning an empty string better than returning nil?

File: .travis.yml (hunk 8 above)
Comment (paween) [reply .travis.yml]: Just created a pull request to separate build changes. https://github.com/iziz/libPhoneNumber-iOS/pull/222

File: libPhoneNumberTests/NBPhoneNumberUtil+ShortNumberTestHelper.m (hunk 14 above)
Comment (paween) [reply libPhoneNumberTests/NBPhoneNumberUtil+ShortNumberTestHelper.m]: This is done to match the Java implementation of the same method.

I think matching the canonical implementation makes porting changes easier.

https://github.com/googlei18n/libphonenumber/blob/master/java/libphonenumber/src/com/google/i18n/phonenumbers/ShortNumberInfo.java#L375

File: libPhoneNumber/NBPhoneNumberUtil+ShortNumber.m (hunk 13 above)
Comment (paween) [reply libPhoneNumber/NBPhoneNumberUtil+ShortNumber.m]: It certainly could. IMHO, there are a couple of reasons that splitting the files that I think outweighs the confusion this might cause:
1. NBPhoneNumberUtil header is not nullability-annotated, but short number support is properly annotated, which means that to add short number support into existing NBPhoneNumberUtil will require its header to be nullability-annotated which will break the code that's currently using NBPhoneNumberUtil in Swift. (I think in a long run, all header files must be properly annotated, but in the meanwhile, I'd like to reduce breakage as much as possible).

2. Since the canonical version of libPhoneNumber (Java) actually separates short number support into its own ShortNumberInfo class, even though it makes sense for us to put these methods as part of NBPhoneNumberUtil, making it into a separate category will make porting changes a lot easier to deal with.

I think since short number support is optional, and if the user wants it, they already have to add a compiler flag, making the user imports a separate file makes sense to me. Though, I think I should also add instructions in README.md to explain how to use short number support.

File: libPhoneNumber/Internal/NBPhoneNumberUtil+Category.h (hunk 12 above)
Comment (paween) [reply libPhoneNumber/Internal/NBPhoneNumberUtil+Category.h]: Removed the file

File: libPhoneNumber/Internal/NBRegExMatcher.h (hunk 11 above)
Comment (paween) [reply libPhoneNumber/Internal/NBRegExMatcher.h]: This is to match the canonical libPhoneNumber (Java) implementation (though the Java version is an interface).

https://github.com/googlei18n/libphonenumber/blob/master/java/libphonenumber/src/com/google/i18n/phonenumbers/internal/MatcherApi.java

When we talked to the libPhoneNumber maintainer at Google about porting (our team is doing the same porting to JavaScript libPhoneNumber), they prefer a direct port. Even though the iOS fork is not maintained by the libPhoneNumber, I feel like we should try to be as close to the canonical code as possible with Objective-C coding code idiom in mind.

In this case, it looks like NBPhoneNumberUtil is doing a bunch of regex stuff that shouldn't be in there, so I can pull this method into NBPhoneNumberUtil if you feel strongly about it.

If we were to pull this into NBPhoneNumberUtil, I'll remove the NBRegularExpressionCache as well.

File: libPhoneNumber/Internal/NBRegularExpressionCache.m (hunk 10 above)
Comment (paween) [reply libPhoneNumber/Internal/NBRegularExpressionCache.m]: I can pull this out into its own PR, though that depends on whether we still want to have NBRegExMatcher. If we don't want the separate NBRegExMatcher, I can remove this class as well.

File: libPhoneNumberTests/NBPhoneNumberUtil+ShortNumberTestHelper.m (hunk 14 above)
Comment (LowAmmo) [reply libPhoneNumberTests/NBPhoneNumberUtil+ShortNumberTestHelper.m]: +1

Understand the oddities...but...I think it makes sense to try to match logic between the Java/Google version and this version...

File: libPhoneNumber/Internal/NBRegExMatcher.h (hunk 11 above)
Comment (dmaclach) [reply libPhoneNumber/Internal/NBRegExMatcher.h]: hmm.. I'll look at trying to clean this up later.

File: libPhoneNumber/Internal/NBRegularExpressionCache.m (hunk 9 above)
Comment (dmaclach) [reply libPhoneNumber/Internal/NBRegularExpressionCache.m]: any comment here?

File: libPhoneNumber/NBMetadataHelper.m [hunk 15]
```diff
@@ -186,4 +193,67 @@ + (BOOL)hasValue:(NSString *)string {
...
+    static NSDictionary *shortNumberDataDictionary;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        // Data is a gzipped JSON file that is embedded in the binary.
+        // See GeneratePhoneNumberHeader.sh and PhoneNumberMetaData.h for details.
+        NSMutableData* gunzippedData =
+            [NSMutableData dataWithLength:kShortNumberMetaDataExpandedLength];
+
+        z_stream zStream;
+        memset(&zStream, 0, sizeof(zStream));
```
Comment (dmaclach) [inline libPhoneNumber/NBMetadataHelper.m]: since this essentially duplicates the other code for reading this stuff, can we move into a single function?

File: libPhoneNumber/NBPhoneNumberUtil+ShortNumber.h [hunk 16]
```diff
@@ -0,0 +1,198 @@
...
+ * <p>
+ * This method takes into account cases where the number might contain formatting, but doesn't
+ * allow additional digits to be appended. Note that {@code isEmergencyNumber(number, region)}
+ * implies {@code connectsToEmergencyNumber(number, region)}.
+ *
+ * @param number the phone number to test
+ * @param regionCode the region where the phone number is being dialed
+ * @return whether the number exactly matches an emergency services number in the given region
+ */
+- (BOOL)isEmergencyNumber:(NSString *)number forRegion:(NSString *)regionCode;
```
Comment (dmaclach) [inline libPhoneNumber/NBPhoneNumberUtil+ShortNumber.h]: why are the arguments on these methods strings instead of NBPhoneNumber?

Comment (paween) [reply libPhoneNumber/NBPhoneNumberUtil+ShortNumber.h]: According to the [canonical implementation](https://github.com/googlei18n/libphonenumber/blob/ed6b1ff4995ff562294aa00cdb718fa7888a0986/java/libphonenumber/src/com/google/i18n/phonenumbers/ShortNumberInfo.java#L431) comment, `isEmergencyNumber` implies `connectsToEmergencyNumber` and it requires a string because it needs to be able to distinguish between cases such as "+1 911" and "911", where some might connect and some might not.

File: libPhoneNumber/Internal/NBRegularExpressionCache.m (hunk 9 above)
Comment (paween) [reply libPhoneNumber/Internal/NBRegularExpressionCache.m]: Done

**Comment by 4tress** [comment]
'NBRegExMatcher.h' file not found after last changes

**Comment by johnryan** [comment]
@4tress did you ever get this working? I tried using the latest master commit but seeing the same issue.

**Comment by 4tress** [comment]
@johnryan we're using previous commit still. Can't say exactly what version it is, I'm on vacations now, will return back to my duties on Monday.
